ocaml/typing/printtyp.ml

```diff
@@ -1300,27 +1289,31 @@ let rec tree_of_typexp mode ty =
           if !print_labels || is_omittable l then outcome_label l
           else Nolabel
         in
+        (* [marg] will contain undetermined axes. It would be imprecise if we
+          don't print anything for those axes, since user would interpret that
+         as legacy. The best we can do is to zap to legacy and if they do land
+         at legacy, we will be able to omit the printing. *)
+        let am = Alloc.zap_to_legacy marg in
```

Comment:
I find this really scary. It means that printing a type can change inference. If we print during debugging, say, we're now altering the course of type inference. I think we shouldn't do this.

Instead, what if we printed e.g. `local?` in this scenario? Obviously that's not parseable. But it's more honest than anything we could say. Or it could be `'_locality` maybe, invoking the weak polymorphism around the value restriction.

Reply:
> If we print during debugging,

Would it be better to use `Printtyped` for debugging? That one is more informative.

If we really need to call `Pprinttyp` for debugging, I suppose `Printtyp` can revert the changes (using log) after printing. 

> Instead, what if we printed e.g. local? in this scenario? 

Very often the root error is argument/return type mismatch rather than mode mismatch, in which case we want to deemphasize the modes. `local?` would be too conspicuous.

Reply:
Huh, so there is actually calling to `Btype.backtrack` in `printtyp.ml` - should we do that here too?

Reply:
Backtracking would be OK.... but sometimes the problem will be the modes, and it seems willfully annoying for us to unconditionally suppress the modes. We could imagine choosing whether or not to print this based on whether or not modes are the issue in the error (the original call into this module likely knows this). Have you tried the `local?` approach? How much does it clutter real output?

And in answer to your first question, I think people (I am one of these people) will want to use pretty-printed typed during debugging.

My guess is that the answer will be to have the original caller say whether or not modes should be printed.

Reply:
I think we should continue the current approach of not printing a mode when it is unconstrained. We already know that the mode in question can’t be the cause of a type error, since that would require it be constrained in some way. I can’t see how printing anything else here won’t just add unhelpful information to a huge percentage of error messages.

We also shouldn’t have type printing mutate anything. That’s bound to cause trouble, as Richard suggests.

Reply:
>  but sometimes the problem will be the modes, and it seems willfully annoying for us to unconditionally suppress the modes.

Yes those cases are also fine. Because our approach here is to suppress modes as much as allowed by the constraints; In the case of mode mismatch, `mode0` and `mode1` will prints differently, even though both are suppressed.

Reply:
OK I've pushed a commit that does backtracking.

Reply:
What if there's an equality check between a mode constrained between `unique` and `exclusive` and one that is known to be `aliased`? I agree the problem can't arise on two-mode axes, but that won't last forever. All that said, it's likely reasonable to delay this until it happens in practice. But when it does, if we suppress this printing, the person who gets the error will likely lose hours of time getting mad at us.

Reply:
The mode constrained between `unique` and `exclusive` will be zapped towards legacy, which is `aliased`. So it will be printed as `exclusive`.

The other one will be printed as `aliased`.

Reply:
My preferred solution for 3-valued axes is not actually to zap to legacy in the printer. Rather I would ensure that the solver, when there is a failure still makes the constraint as true as possible. So if `m` cannot be made less than `n` it should zap `m` to its lower bound and zap `n` to its upper bound. That way any modes causing an error are constant and the printer will show exactly the useful information.

This follows a general principle that constraint solvers should make constraints as true as they can, so that error printing highlights only the genuine differences.

Zapping to legacy will sometimes have this effect — as in your example — but will sometimes go the other way and make modes look further apart than they really are. So I’d rather the printer just continued to ignore non-constant modes.

Once we have the unzipped syntax, I’d like to go further and try avoiding printing any modes when the error is about types, and any types when the error is about modes. Not sure if that will work for every error, but I think it’s the right direction to go.

Reply:
To rephrase what you said:
- When submoding fails in `ctype.ml`, we should make the modes into constants such that they are close to each other.
- In printer, we still want to call `zap_to_legacy` on modes, because there will be undetermined modes, and we want to de-emphasize those.

Yes I agree with that. I wonder at which level should we implement "leave submode error in the closest state". I looked at `solver.ml` briefly, and it would be too much effort to re-work the pre/post-conditions of those mutually recursive functions.  Considering that the errors are not on the production path so performance doesn't matter, it's easiest to wrap each call to `submode/equate` in `ctype.ml` with a further `submode` that does the constraint in case of error. This is easy because `submode a b` already gives in the case of error `{left; right}` where `left` is the lowest `a` can be, and `right` is the highest `b` can be. 

> So I’d rather the printer just continued to ignore non-constant modes.

We can't just ignore non-constant modes - a mode constrained between `unique` and `exclusive` still needs to be printed, for otherwise the user would assume `shared`.

Reply:
It should be in `solver.ml` on the error path. It is a concern of the mode solver not `ctype.ml`. You shouldn’t need to change any recursive functions, just add another layer of wrapping if you need it. The old solver already needed a layer of wrapping for converting exceptions to results, I would put this code in there if it still exists.

> We can't just ignore non-constant modes - a mode constrained between unique and exclusive still needs to be printed

I disagree. Whatever you print there will confuse people because it either won’t be accurate or it won’t be valid syntax. If you print nothing users just won’t see it and we already no it is not the part of the type the user needs to see because it’s not relevant to the error.

I think the right model is not: “`a -> b` says that the function takes a global value and returns a global value”, but instead “`a -> b` does not specify what modes the argument and return value have. Combined with a rule that says if we don’t know the mode for something we default it to legacy. I think this model makes what is going on with the unzipped syntax much clearer. It also allows error messages to just leave out modes that aren’t important (just as the type printer should, but currently doesn’t, replace unimportant parts of types with `_`).

Perhaps it would help if we changed the meaning of `(e : a -> b)` to not constrain the modes of the function. So users would write `( e : a @ . -> b @ .)` if they really wanted to constrain the modes.

> we still want to call zap_to_legacy on modes

For clarity, I would prefer to not print such modes rather than zap them to legacy. I don’t mind zapping them to legacy for now, since for two-mode axes that’s very close to the same thing.

Reply:
Yes, if we say `a -> b` doesn't specify any modes at all, then I'm happy to omit printing modes when they are not the root error. BTW this probably needs overhauling `ctype.ml` to pinpoint the exact mismatched mode; but I guess that will be subsumed by the PR of unzipped syntax.

In signatures `a -> b` will still default to legacy.

I also wonder how much of user's code relies on `a -> b` meaning legacy on LHS&RHS. I guess it should be mostly fine, because we are making more programs type check.